h2o-core/src/main/java/hex/grid/HyperSpaceWalker.java

```diff
@@ -70,7 +72,51 @@
      *
      * @return  array of "untyped" values representing configuration of grid parameters
      */
-    Object[] getCurrentRawParameters();
+    abstract public Object[] getCurrentRawParameters();
+
+    /**
+     * 
+     * @param permutation single permutation from the grid, represented as array of Objects and ordered based on {@code hyperParamNames} array
+     * @param hyperParamNames array of hyper parameters names grid search is performed over
+     * @param filterFunctions list of {@link PermutationFilterFunction} to check against. 
+     *                        Permutation will be skipped if any of the functions consider it as to be skipped one.
+     */
+    boolean permutationIsSkipped(Object[] permutation, String[] hyperParamNames, ArrayList<PermutationFilterFunction> filterFunctions) {
```

Comment:
this just does not belong here

Reply:
@honzasterba could you please elaborate on this one? 
I will try to give more reasoning here from my side. For example current implementation of `RandomDiscreteValueWalker` has `nextModelIndices` methods that actually skips already visited entries. This method actually says where walker should not step. In the same fashion `permutationIsSkipped` is controlling where we should not step, but filtering logic could be generalised and reused by all `HyperSpaceWalker`'s implementations. 
I also did not want to change interface to abstract class but it is fine for interface to have implementations. Java 8 provides `default` keyword to support this direction. I just did not want to highlight that this is a default behaviour and user should consider to override it. So I would say we are experiencing lack of expressiveness of the language here.

Reply:
bottom line - the hyperspacewalker does enough as it is, if it can at all avoided new functionality should be added by extension, either by composition or inharitance, in this case composition is the best way to go (and the more flexible one)
`nextModelIndices` method is completely unrealted to filtering, it only prevets the same params to be emitted twice

Reply:
> nextModelIndices method is completely unrealted to filtering, it only prevets the same params to be emitted twice

it depends to what extend you stretch an abstraction. First match filtering could be seen as ' prevent the same(semantically) params to be emitted twice` as well.

Reply:
yes it could be, if you are fine with your program being stuck in an infinite loop

Reply:
same as training unnecessary permutations without filtering. It is all about saving time by avoiding permutations that we don't need( either duplicates or semantically the same permutations)

Reply:
IMHO program correctness (e.g. ever reaching a result) and program time complexity are quite distinct concepts that should not be mistaken for one another. If you don't see a difference in such concepts then I am afraid we are not getting anywhere with this.
To paraphrase, the `nextModelIndices` is necessary for the program ever reach a result, since moving it outside of the Walker, the wakler-filter duo would be stuck in an infinite look, therefore the program would be incorrect (non-returning). 
On the other hand, filtering out unwanted permutation is a feature thats separate from finitely walking the hyperspace and such such its more of an optimisation that can be external to the responsibilty of finding a path through the hyperspace.

Reply:
sounds like what @honzasterba is saying would result in a better separation of concerns. WDYT @deil87?

Reply:
@honzasterba I do see the difference that you are trying to explain me. Please try to understand my point of view. Without logic that prevents duplicates in “nextModelIndices” program could be considered a correct one - it can finish once we visited all unique permutations(visiting duplicates would just spend our time). Filtering is the same from that perspective. So it is all could be seen as just about time complexity.